### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by Karyum**
rel #37 

added failed tests, refactored code ( spiced it up with some async/await ).
added .istanbul.yml to ignore the files that can't be tested ( for example oauth, and posting to the api ).
setup codecov.
**100% coverage**

sorry for the big pr  

ps: some dev dep look like they are reinstalled cause i had problems with my npm :sweat_smile:

### CODE REVIEW

File: routes/event_details.js [hunk 1]
```diff
@@ -4,11 +4,11 @@ module.exports = (req, res) => {
   var url = 'https://nazareth-open-tourism-platform.herokuapp.com/events/' + id;
 
   request.get(url, (error, result) => {
-    var event = JSON.parse(result.body);
-
     if (error) {
       res.send('error :(');
     } else {
+      console.log(1);
```
Comment (mattlub) [inline routes/event_details.js]: ?

File: tests/tests.js [hunk 2]
```diff
@@ -1,14 +1,42 @@
...
       t.equal(res.status, 200, 'should return status code 200');
       t.ok(res.text.includes('<div id="list-page-content">'), 'should return home page containing test string "<div id="list-page-content">"');
       t.end();
     });
 });
+
+tape('Test the event details woute', t => {
+  const htmlSample = '<h1 class="text-warning eventName">default</h1>';
+  supertest(server)
+    .get('/events/59afaf75b0a5e80011beeafa')
```
Comment (mattlub) [inline tests/tests.js]: where is this id coming from?

File: tests/tests.js [hunk 3]
```diff
@@ -1,14 +1,42 @@
...
     .end((err, res) => {
-      if (err) throw err;
+      t.error(err, 'No Erro');
       t.equal(res.status, 200, 'should return status code 200');
       t.ok(res.text.includes('<div id="list-page-content">'), 'should return home page containing test string "<div id="list-page-content">"');
       t.end();
     });
 });
+
+tape('Test the event details woute', t => {
```
Comment (mattlub) [inline tests/tests.js]: 'route' is spelled wrong

File: tests/tests.js [hunk 4]
```diff
@@ -1,14 +1,42 @@
...
 const server = require('../server.js');
+const request = require('request');
 
 tape('home route test: GET request to /', t => {
-  request(server)
+  supertest(server)
     .get('/')
     .end((err, res) => {
-      if (err) throw err;
+      t.error(err, 'No Erro');
```
Comment (mattlub) [inline tests/tests.js]: you don't need to provide a message here:
see https://www.npmjs.com/package/tape#terrorerr-msg

File: tests/tests.js (hunk 2 above)
Comment (Karyum) [reply tests/tests.js]: I added it manually and now thinking about it it's probably bad practice.
whats the better way to do it ?

File: tests/tests.js (hunk 4 above)
Comment (Karyum) [reply tests/tests.js]: is it preferred to use like an if statement and t.fail after it in case of an error  ?

File: routes/event_details.js (hunk 1 above)
Comment (Karyum) [reply routes/event_details.js]: yeah forgot about that :stuck_out_tongue_winking_eye: cheers

File: tests/tests.js (hunk 4 above)
Comment (mattlub) [reply tests/tests.js]: just `t.error(err)` is easier

File: tests/tests.js (hunk 2 above)
Comment (mattlub) [reply tests/tests.js]: I don't know 😕

Comment (mattlub) [reply tests/tests.js]: you could query the API first to get all events, then use one of the id's you get back?

But this is a bit long

Comment (Karyum) [reply tests/tests.js]: yeah but that sounds better, i'll just make a GET to the api take the first object and use it to test

**Comment by Karyum** [comment]
@JWLD @mattlub not sure who needs to QA this but it's ready :smile:

**Comment by JWLD** [comment]
Tests are currently failing @Karyum @MohamedOmarii

File: tests/tests.js [hunk 5]
```diff
@@ -1,14 +1,77 @@
...
       t.end();
     });
 });
+
+tape('Test the event details route', t => {
+  const htmlSample = '<h1 class="text-warning eventName">FACN2</h1>';
+  let eventId = '';
+  request.get('https://nazareth-open-tourism-platform.herokuapp.com/events', (err, res) => {
+    if (err) return err;
+    const filteredResult = JSON.parse(res.body).filter((event) => event.en.name === 'FACN2');
```
Comment (mattlub) [inline tests/tests.js]: what if there is no event with name FACN2?
you don't need to get a specific event's id, just any event id.

File: tests/tests.js [hunk 6]
```diff
@@ -1,14 +1,77 @@
...
+      .expect(200)
+      .end((err, res) => {
+        t.error(err);
+        t.ok(res.text.includes(htmlSample), 'It found the right event');
+        t.end();
+      });
+  });
+});
+
+// Work in progress
```
Comment (mattlub) [inline tests/tests.js]: finish this or delete it?

File: tests/tests.js [hunk 7]
```diff
@@ -1,14 +1,77 @@
...
+    .get('/add-event')
+    .expect(200)
+    .end((err, res) => {
+      t.error(err);
+      t.ok(res.text.includes(htmlSample));
+      t.end();
+    });
+});
+
+tape('Test if it posts the event to the api', t => {
```
Comment (mattlub) [inline tests/tests.js]: we can't have these tests actually posting data to the API.

Comment (Karyum) [reply tests/tests.js]: So should i not test POST requests?

**Comment by Karyum** [comment]
@mattlub I'm not sure what to test next any idea?
frontend maybe?

**Comment by mattlub** [comment]
@karyum test the authentication middleware?

File: tests/tests.js (hunk 7 above)
Comment (mattlub) [reply tests/tests.js]: You shouldn't actually send the post requests to the platform, because it will add actually data to the production database (and we have tests on the platform to check it works correctly anyway). You could do something like what Elias showed you if you want to test your routes without actually sending the post

**Comment by Karyum** [comment]
@mattlub done. anything more??

File: tests/tests.js [hunk 8]
```diff
@@ -33,3 +33,21 @@ tape('Test the event details route', t => {
       });
   });
 });
+
+tape('Test the authentication middleware', t => {
+  supertest(server)
+    .get('/add-event')
+    .set('Cookie', [ 'token=eyJhbGciOiJIUzI1NiJ9.NTdhYjM2NTQ0NGUxYjhhOTY3NzExMWMzZjYxY2VjNzFiNzMyNjZhYw.Fpzao_FN05j29auCfrrM0-eNhqYJQ_QYbzJiXnZXSQA' ])
```
Comment (mattlub) [inline tests/tests.js]: you shouldn't hard code this, use `jwt.sign`

**Comment by mattlub** [comment]
@Karyum I would say if there's nothing more that's obvious for you to test then leave it, some more tests would be better but probably not the priority right now?

**Comment by Karyum** [comment]
@mattlub whats left to test is the oauth which i think can't be tested, and posting to the api which i also can't do.

**Comment by Karyum** [comment]
@mattlub travis fixed

File: tests/tests.js (hunk 8 above)
Comment (mattlub) [reply tests/tests.js]: ^^

Comment (mattlub) [reply tests/tests.js]: @Karyum

**Comment by Karyum** [comment]
@mattlub i guess you could merge it since @JWLD didn't touch much on supertest

File: tests/tests.js (hunk 8 above)
Comment (Karyum) [reply tests/tests.js]: @mattlub sorry about didn't see it i was trying to fix travis lol

Comment (Karyum) [reply tests/tests.js]: @mattlub :shipit:

File: tests/tests.js [hunk 9]
```diff
@@ -1,14 +1,79 @@
...
+  }
+});
+
+tape('Test the authentication middleware', t => {
+  const token = jwt.sign('somthing', process.env.JWT_SECRET);
+  supertest(server)
+    .get('/add-event')
+    .set('Cookie', [`token=${token}`])
+    .end((err, res) => {
+      t.error(err);
```
Comment (finnhodgkin) [inline tests/tests.js]: In my opinion tests should be descriptive even when they pass. Something like `/add-event route doesn't return an error`

Seeing a bunch of `null` test messages isn't informative. `t.error` does display the error message when there's an error, but that message may not be informative when it happens, and anyway doesn't _not_ display the error message when you do supply a test description.

File: tests/tests.js [hunk 10]
```diff
@@ -1,14 +1,79 @@
...
+      t.ok(
+        res.text.includes(
+          '<form class="add-event-form" action="/add-event" method="post">'
+        )
+      );
+      supertest(server)
```
Comment (finnhodgkin) [inline tests/tests.js]: Why is this `supertest` nested inside the last?

File: tests/tests.js [hunk 11]
```diff
@@ -1,14 +1,79 @@
...
+      });
+    supertest(server)
+      .get('/en/events/someid')
+      .end((err, res) => {
+        t.error(err);
+        t.ok(res.text.includes('<title>Not Found</title>'));
+        t.end();
+      });
+  } catch (err) {
+    console.log(err);
```
Comment (finnhodgkin) [inline tests/tests.js]: Try catch shouldn't really be used within tests. If you want to test if something within your tests throws an error, use [t.throws](https://github.com/substack/tape#tthrowsfn-expected-msg) or similar

Comment (finnhodgkin) [reply tests/tests.js]: Also `console.log`s aren't really tests...

File: tests/tests.js [hunk 12]
```diff
@@ -1,14 +1,79 @@
...
+      t.error(err);
       t.equal(res.status, 200, 'should return status code 200');
-      t.ok(res.text.includes('<div id="list-page-content">'), 'should return home page containing test string "<div id="list-page-content">"');
+      t.ok(res.text.includes('<div id="list-page-content">'));
+      t.end();
+    });
+});
+
+tape('Test the event details route', async t => {
+  let eventId = '';
```
Comment (finnhodgkin) [inline tests/tests.js]: Why is this a let and not just declared below?

File: routes/event_details.js [hunk 13]
```diff
@@ -1,19 +1,22 @@
...
-      res.render('event_details', {
-        title: 'Event Details',
-        eventText: JSON.parse(result.body)[req.params.lang],
-        event: JSON.parse(result.body),
-        place: JSON.parse(result.body).place[req.params.lang],
-        back: req.headers.referer
-      });
-    }
-  });
+module.exports = async (req, res) => {
```
Comment (finnhodgkin) [inline routes/event_details.js]: Keep PRs focused. If the whole code-base is moving to async await then make a PR dedicated to just that, rather than pushing it bit by bit with other PRs. :+1:

File: tests/tests.js [hunk 14]
```diff
@@ -1,14 +1,79 @@
...
+  try {
+    const eventResponse = await axios.get(
+      'https://nazareth-open-tourism-platform.herokuapp.com/api/v1/events'
```
Comment (finnhodgkin) [inline tests/tests.js]: You should probably have two environments, a dev setup that's running a local version of the OTP and a production one that's linked to the real OTP. That would fix two important problems:

1. Not being able to mess around with the production database (add dummy events within tests, etc.)
1. Having to wait x seconds for your tests to run.

File: tests/tests.js (hunk 11 above)
Comment (Karyum) [reply tests/tests.js]: yeah forgot about that good spot :+1:

**Comment by Karyum** [comment]
@finnhodgkin changes done. ready for review :kissing_heart:

**Comment by finnhodgkin** [comment]
As discussed in person, use nock to mock all the calls to the OTP so the tests pass on Travis and can be run locally without an internet connection/the OTP.

File: tests/flush_database.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+  mongoose.connect(dbURI, {
+    useMongoClient: true
+  });
+
+  mongoose.connection.on('open', () => {
+    mongoose.connection.db.dropDatabase(function (err, res) {
+      console.log(err);
+      console.log(res);
+      process.exit(0);
```
Comment (finnhodgkin) [inline tests/flush_database.js]: If you must have console.logs in here at least make it something like `console.log(err || res)` :stuck_out_tongue_winking_eye:

File: tests/tests.js [hunk 16]
```diff
@@ -1,14 +1,184 @@
...
+const supertest = require('supertest');
 const server = require('../server.js');
+const axios = require('axios');
+const jwt = require('jsonwebtoken');
+const nock = require('nock');
+const flush = require('./flush_database.js');
+require('dotenv').config();
+
+let id;
+axios
```
Comment (finnhodgkin) [inline tests/tests.js]: This should be nocked. No requests in the tests.

File: tests/tests.js [hunk 17]
```diff
@@ -1,14 +1,184 @@
...
 tape('home route test: GET request to /', t => {
-  request(server)
+  supertest(server)
     .get('/')
     .end((err, res) => {
-      if (err) throw err;
+      t.error(err, '/ home route does not return an error');
       t.equal(res.status, 200, 'should return status code 200');
-      t.ok(res.text.includes('<div id="list-page-content">'), 'should return home page containing test string "<div id="list-page-content">"');
+      t.ok(res.text.includes('<div id="list-page-content">'));
```
Comment (finnhodgkin) [inline tests/tests.js]: Always test messages.

File: tests/tests.js [hunk 18]
```diff
@@ -1,14 +1,184 @@
...
+      t.end();
+    });
+});
+
+tape('Test POST arabic event', t => {
```
Comment (finnhodgkin) [inline tests/tests.js]: Not a blocker but ideally I'd like to see a bit of edge case handling. What happens when you send bad data or missing properties, etc.?

File: tests/tests.js [hunk 19]
```diff
@@ -1,14 +1,184 @@
...
+  nock(process.env.URI)
+    .post('/places')
+    .reply(200, { _id: 1 });
+
+  supertest(server)
+    .post('/add-event')
+    .send(data)
+    .end((err, res) => {
+      t.error(err, 'posting an event did not error');
+      console.log(res.statusCode);
```
Comment (finnhodgkin) [inline tests/tests.js]: No `console.log`s in the tests. You could replace it with a test for statusCode 200 :man_shrugging:

File: routes/middleware/post_place.js [hunk 20]
```diff
@@ -1,7 +1,8 @@
 const request = require('request');
+require('dotenv').config();
 
 module.exports = (req, res, next) => {
```
Comment (finnhodgkin) [inline routes/middleware/post_place.js]: This middleware should only run if the user hasn't selected a place that already exists

File: routes/middleware/post_place.js [hunk 21]
```diff
@@ -24,8 +25,11 @@ module.exports = (req, res, next) => {
   };
 
   request(options, (error, result, body) => {
+    /* istanbul ignore next */
     if (error) {
-      res.send(error);
+      res.render('home', {
+        title: 'Not found'
```
Comment (finnhodgkin) [inline routes/middleware/post_place.js]: You should probably add a log to all these errors so you can at least tell how things are going wrong on the backend.

File: tests/tests.js [hunk 22]
```diff
@@ -1,14 +1,184 @@
...
+    endTime: '14:00',
+    cost: 0,
+    placeId: id
+  };
+
+  nock(process.env.URI)
+    .post('/events')
+    .reply(200, { _id: 1 });
+
+  nock(process.env.URI)
```
Comment (finnhodgkin) [inline tests/tests.js]: If you make the above change to `post_place` you won't need to nock this when it's a form selected place.

File: routes/middleware/get_places.js [hunk 23]
```diff
@@ -1,36 +1,27 @@
-const request = require('request');
+const axios = require('axios');
 
-module.exports = (req, res, next) => {
-  const url = 'https://nazareth-open-tourism-platform.herokuapp.com/api/v1/places';
+module.exports = async (req, res, next) => {
+  const url = `${process.env.URI}/places`;
+  const placesResponse = await axios.get(url);
+  res.locals.places = placesResponse.data
+    .map(place => {
```
Comment (finnhodgkin) [inline routes/middleware/get_places.js]: I would say maybe this map/filter should be moved to a helper function so you can test it separately.

File: tests/tests.js [hunk 24]
```diff
@@ -1,14 +1,184 @@
...
+    });
+});
+
+tape('Test the authentication middleware', t => {
+  nock(process.env.URI)
+    .get('/places')
+    .reply(200, [
+      { ar: { name: 'سيس' } },
+      { en: { name: 'base' } },
+      { en: { name: 'home' } }
```
Comment (finnhodgkin) [inline tests/tests.js]: If you changed the map/sort in `get_places` to a helper function then you could get rid of this section and just test the function itself, rather than the route.

File: views/add_event.hbs [hunk 25]
```diff
@@ -199,7 +199,7 @@
         <label>Place </label>
         <div id="addPlaceIcon" class="fa fa-plus-circle fa-2x add-place-icon"></div>
       </div>
-      <select id="selectAllPlaces" class="place-options " name="placeId" required>
+      <select id="selectAllPlaces" class="place-options" name="placeId">
         <option disabled selected></option>
         {{#each places}}
           <option class="placesList" value="{{this.id}}">{{this.name}}</option>
```
Comment (finnhodgkin) [inline views/add_event.hbs]: What's up with this indenting?

File: public/js/get_events.js [hunk 26]
```diff
@@ -1,8 +1,12 @@
 /* global XMLHttpRequest */
 var getEventsModule = function (queryParams, callback) {
-  var baseUrl = 'https://nazareth-open-tourism-platform.herokuapp.com/api/v1/events';
+  var baseUrl = 'http://localhost:3000/api/v1/events';
```
[the remark this answers is not in the document]
Comment (Karyum) [reply public/js/get_events.js]: can't do process env since this is the frontend but i'll change it back to the actual api

[the remark this answers is not in the document]
Comment (finnhodgkin) [reply public/js/get_events.js]: Oops yeah, actual api or the production app won't work...

File: public/js/get_events.js [hunk 27]
```diff
@@ -1,8 +1,12 @@
...
   if (queryParams.date_from && queryParams.date_to) {
-    baseUrl += '/?date_from=' + queryParams.date_from + '&date_to=' + queryParams.date_to;
+    baseUrl +=
```
[the remark this answers is not in the document]
Comment (finnhodgkin) [reply public/js/get_events.js]: Sorry didn't realise this was front-end. Ignore this.

File: routes/middleware/authenticate.js [hunk 28]
```diff
@@ -2,23 +2,32 @@ const qs = require('querystring');
 require('dotenv').config();
 const jwt = require('jsonwebtoken');
 
-module.exports = (req, res, next) => {
+module.exports = async (req, res, next) => {
   const tokenCookie = req.cookies.token;
   const queries = qs.stringify({
```
[the remark this answers is not in the document]
Comment (Karyum) [reply routes/middleware/authenticate.js]: :arrow_up:

File: routes/middleware/get_places.js [hunk 29]
```diff
@@ -1,36 +1,27 @@
-const request = require('request');
+const axios = require('axios');
 
-module.exports = (req, res, next) => {
-  const url = 'https://nazareth-open-tourism-platform.herokuapp.com/api/v1/places';
+module.exports = async (req, res, next) => {
+  const url = `${process.env.URI}/places`;
+  const placesResponse = await axios.get(url);
```
[the remark this answers is not in the document]
Comment (Karyum) [reply routes/middleware/get_places.js]: and what do i do for error handling should i redirect the user to somewhere else orr??? 

cause for now i'm adding a next() so that would show an empty dropdown list :grimacing: any idea?

[the remark this answers is not in the document]
Comment (finnhodgkin) [reply routes/middleware/get_places.js]: Hmmm.. There should be some kind of error page I guess :man_shrugging: The only time I can imagine it happening is if the OTP is down, in which case nothing else on NE will work.

File: routes/middleware/post_place.js (hunk 21 above)
Comment (Karyum) [reply routes/middleware/post_place.js]: so i guess only in development im allowed to have console logs in the backend right ?

Comment (Karyum) [reply routes/middleware/post_place.js]: but im testing for errors in my tests so my terminal would just get filled up with errors, is it still a good idea?

File: tests/tests.js (hunk 24 above)
Comment (Karyum) [reply tests/tests.js]: still need to test the route that runs the function

File: views/add_event.hbs (hunk 25 above)
Comment (Karyum) [reply views/add_event.hbs]: that's actually correct indentation the first option is empty and then im getting the rest of the places

File: tests/tests.js (hunk 18 above)
Comment (Karyum) [reply tests/tests.js]: not sure if i can test that since we are mocking the post request which is giving back a valid id.
so no matter what it's going to work.
also there is "required" on properties that the api must have in order for a successful POST request.

ps: #75

**Comment by Karyum** [comment]
@finnhodgkin i called the folder for the sort function helpers, but im not sure about it any ideas?

File: routes/middleware/post_place.js (hunk 21 above)
Comment (finnhodgkin) [reply routes/middleware/post_place.js]: I was actually thinking more along the lines of it being _good_ to have `console.log`s in production so there's a record of any errors.

File: tests/tests.js (hunk 18 above)
Comment (finnhodgkin) [reply tests/tests.js]: Yes, but I'm guessing the OTP sends back some sort of error message when it recieves a bad request. You could nock the bad response and test that you have appropriate error handling for it.

**Comment by finnhodgkin** [comment]
Sounds like a good name to me @Karyum :+1:

File: routes/middleware/post_place.js (hunk 21 above)
Comment (finnhodgkin) [reply routes/middleware/post_place.js]: Not a blocker or particularly relevant to this PR...
